ssas/systems.go

```diff
@@ -414,6 +431,41 @@ func GenerateSecret() (string, error) {
 	return fmt.Sprintf("%x", b), nil
 }
 
+/*
+	ResetSecret creates a new secret for the current system.
+*/
+func (system *System) ResetSecret(trackingID string) (string, error) {
+	db := GetGORMDbConnection()
+	defer Close(db)
+
+	newSecretEvent := Event{Op: "ResetSecret", TrackingID: trackingID, ClientID: system.ClientID}
+	OperationStarted(newSecretEvent)
+
+	secretString, err := GenerateSecret()
+	if err != nil {
+		newSecretEvent.Help = fmt.Sprintf("could not reset secret for clientID %s: %s", system.ClientID, err.Error())
```

Comment:
Presumably an error here or elsewhere in this function would have a message clear enough to distinguish what failed.  Do you think those that follow us might ever ask, "Did that unique cryptographic error bubble up from `GenerateSecret()` or `NewHash()`?"  I've often come down on the side of generating more unique error messages to combat that possibility, but what you have here has the upside of consistency.

Reply:
The GenerateSecret() and NewHash() methods log their operation start, failure and success, so there should be no wonder here. In contrast to most approaches with error handling, which take a top down approach to logging (log the error at the last place it's handled), we are logging from the bottom up. We're doing this to provide useful diagnostic logs in production and also useful warning metrics.